Name       : Brion Mallon

C A 1      : 

Input      : Use a songs metadata file contains a 10,000 record subset of http://labrosa.ee.columbia.edu/millionsong/
             List all files here

Process    : Apply the following actions 

           - Review the csv files for Data and Web mining Ca 

Songs file : this file contains 10,000 entries with track, artist, year and other attributes  


Start By importing required packages, get input songs file 

In [1]:
%matplotlib inline
import matplotlib as plt
import pandas as pd
import sqlite3 as db
from pandas.io import sql
# Create your connection.
cnx = db.connect(':memory:')    # this ensure sqlite tables stay in memory

##### There are currently 4 metafiles , read these for examination 


###### Files 
            -- songs.csv
            -- artist_tag.csv  
            -- artist_location.csv  
            -- artist_artist_similarty.csv 
 
 ###### Steps
 1. read all files into Dataframes 
 2. Copy the dataframes into SQLlite3 as tables, from here they can be examined using SQL
 3. If required, save any changes and rewrite csv files
       


First step ,  read all csv files 

In [2]:
songs = pd.read_csv("songs.csv", encoding='ISO-8859-1', dtype={'artist_familiarity': float,'artist_hotttnesss': float, 'year': int})  

artist_tag = pd.read_csv("artist_tag.csv" , encoding='ISO-8859-1') 

artist_location = pd.read_csv("artist_location.csv"  ,encoding='ISO-8859-1' , dtype={'Latitude': float,'Longitude': float})  

artist_artist_similarty = pd.read_csv("artist_artist_similarty.csv" , encoding='ISO-8859-1') 


Now, add all above file to a sqllite database

In [3]:
# write out all Dataframes, these can subsequently be read using SQL

sql.to_sql(songs, name='songs', if_exists='replace', con=cnx)
sql.to_sql(artist_tag, name='artist_tag', if_exists='replace', con=cnx)
sql.to_sql(artist_location, name='artist_location', if_exists='replace', con=cnx)
sql.to_sql(artist_artist_similarty, name='artist_artist_similarty', if_exists='replace', con=cnx)

Get a count of rows for each table 

In [4]:
 print(sql.read_sql_query('select count(*) as "songs count" from songs', cnx))
 print(sql.read_sql_query('select count(*) as "artist_tag count" from artist_tag', cnx))
 print(sql.read_sql_query('select count(*) as "artist_location count" from artist_location', cnx))
 print(sql.read_sql_query('select count(*) as "artist_artist_similarty count" from artist_artist_similarty', cnx))

   songs count
0        10000
   artist_tag count
0              3809
   artist_location count
0                   1402
   artist_artist_similarty count
0                          42969


So, the counts look ok. First check the songs table, select sample rows 

In [29]:
q1='select * from songs  limit 10'
pd = (sql.read_sql_query(q1, cnx))
pd.head(10)

index            track_id  \
0      0  TRACCVZ128F4291A8A   
1      1  TRACCMH128F428E4CD   
2      2  TRACCSW128F148C7C3   
3      3  TRACCXJ128F428F0CF   
4      4  TRACCVS12903D014F8   
5      5  TRACCKS128F42B77AE   
6      6  TRACCQM12903CACC1E   
7      7  TRACCUS128F92E1FEB   
8      8  TRACCJA128F149A144   
9      9  TRACCGB12903CD1B90   

                                               title             song_id  \
0                                  Deep Sea Creature  SOVLGJY12A8C13FBED   
1                               No Quieras Marcharte  SOGDQZK12A8C13F37C   
2                                            If I...  SODMVJR12A6D4F985D   
3  Werther - Lyric Drama in four Acts/Act I/Alors...  SOIWBDR12A8C13A4AC   
4               Ad Te Clamamus Exsvles Mortva Liberi  SOHCCIA12AC907577F   
5                                         Murder One  SOBOAQC12A8C13E3E9   
6                                   On My Feet Again  SOKVLHX12AB0187B39   
7                                  Bedroom Acoustics  SOMMSMW12A8C13FCCC   
8                                            Segredo  SODPNJR12A6D4FA52D   
9                          Sajana (Ft Faheem Mazhar)  SOFFLLP12AB018ED52   

                   release           artist_id  \
0     Call of the Mastodon  ARMQHX71187B9890D3   
1                 Adelante  AR2PT4M1187FB55B1A   
2                Ill Na Na  ARDI88R1187B98DAB2   
3        Massenet: Werther  ARUKJUP12086C14589   
4           Pentagrammaton  ARZEWUR1187FB53DC8   
5         BTNHRESURRECTION  ARUZRFN1187B98AC05   
6                   Utopia  ARHBWOZ1187FB3FD53   
7             Plug In Baby  ARR3ONV1187B9A2F59   
8       Joao Voz E Violato  AR3THYK1187B999F1F   
9  The Lost Souls Bonus EP  ARFELOH1187B991F95   

                            artist_mbid                      artist_name  \
0  bc5e2ad6-0a4a-4d90-b911-e9a7e6861727                         Mastodon   
1  d54ea4a6-0b9c-4e47-bed0-289ae9ff4037                      Los Chichos   
2  fd87374e-ffde-4d36-89a8-8a073f795666                       Foxy Brown   
3  8a5f2736-bcde-4a2e-8d50-72631d66a7ef  Ramón Vargas;Vladimir Jurowski   
4  0be59867-0da4-4e45-9b64-728cdf25487c                        Enthroned   
5  2fa45bbb-0efb-4950-9d40-94bf23cbfec1             Bone Thugs-N-Harmony   
6  e6ff2839-5ccb-451b-b07e-f485bc143118                           Utopia   
7  fd857293-5ab8-40de-b29e-55a69d4e4d0f                             Muse   
8  286ec4c2-b5ca-4f85-b331-280a6d73dd14                   João Gilberto   
9  65b785d9-499f-48e6-9063-3a1fd1bd488d                       Niraj Chag   

    duration  artist_familiarity  artist_hotttnesss  year  
0  280.21506            0.780462           0.574275  2001  
1  191.68608            0.561590           0.420570  1984  
2  222.92853            0.687687           0.406686     0  
3  278.38649            0.391741           0.291265     0  
4   89.15546            0.593416           0.395710  2010  
5  255.73832            0.815923           0.555138  2000  
6  239.59465            0.543937           0.430300     0  
7  156.96934            0.929030           0.750427     0  
8  197.19791            0.645192           0.471224  2000  
9  262.58240            0.495819           0.342765     0

In [6]:
# Get a count of unique artists on the songs file
#
q1='select count(distinct(artist_id || artist_name )) from songs'
pd = (sql.read_sql_query(q1, cnx))
pd.head()

count(distinct(artist_id || artist_name ))
0                                        4437

AS there are only 4437 artists for the 10,000 songs, some artists have multiple entries.  we will have to make a seperate file of distinct artists. Select artist id and name .


In [7]:
artist_id_name = (sql.read_sql_query('select distinct artist_id, artist_name from songs', cnx))

# write a version to sqllite
sql.to_sql(artist_id_name, name='artist_id_name', if_exists='replace', con=cnx)

# write a version as a csv file 
artist_id_name.to_csv('artist_id_name.csv',  encoding='ISO-8859-1')

print(len(artist_id_name.index))

4437


 Now, go to the artist_tags table and get a distinct list of tags (first select from table)

In [28]:
q1='select * from artist_tag  limit 10'
pd = (sql.read_sql_query(q1, cnx))
pd.head(10)

index           artist_id             mbtag
0      0  AR00A6H1187FB5402A           detroit
1      1  AR00A6H1187FB5402A              punk
2      2  AR00A6H1187FB5402A          michigan
3      3  AR00A6H1187FB5402A               usa
4      4  AR01VU31187B997DA0         hard rock
5      5  AR01VU31187B997DA0              rock
6      6  AR01VU31187B997DA0          american
7      7  AR01W2D1187FB5912F     united states
8      8  AR022JO1187B99587B  production music
9      9  AR02IU11187FB513F2              jazz

Now, check how many unique Tags exist on on_artist_tag

First, get a list of distinct tags

In [9]:
q1='select distinct(mbtag) from artist_tag'
mbtags = (sql.read_sql_query(q1, cnx))
print(len(mbtags.index))
mbtags.head(20)

726


mbtag
0             detroit
1                punk
2            michigan
3                 usa
4           hard rock
5                rock
6            american
7       united states
8    production music
9                jazz
10              latin
11             trance
12         electronic
13  progressive house
14           composer
15      film composer
16                pop
17          urban pop
18              dance
19      pop and chart

There are 720 distinct mbtags,  create a file of these 

In [10]:
# write a version to sqllite
sql.to_sql(mbtags, name='mbtags', if_exists='replace', con=cnx)
# write a version as a csv file 
mbtags.to_csv('mbtags.csv', encoding='ISO-8859-1')

Now, get count of tags for each artist, order desc by count

In [11]:
q1='select artist_id, count(*) as "count" from artist_tag group by artist_id order by 2 desc'
pd = (sql.read_sql_query(q1, cnx))
pd.head(10) 

artist_id  count
0  AR4L4WQ1187FB51996     20
1  AR6XZ861187FB4CECD     20
2  ARMJMBJ1187B98C46F     20
3  ARXPPEY1187FB51DF4     20
4  ARAXU5Z1187B98FECD     17
5  ART5MUE1187B98C961     17
6  AR5FP401187FB523C9     16
7  ARASYMJ1187B9ACAF2     16
8  ARN5NHX1187FB4F755     16
9  ARNHHIV1187FB4EA23     16

Now, get a count of the number of artists have a given number of tags  (e.g say 3 artists have 20 and so on)
( do an inner and outer group by )

In [59]:
q1 = 'select a.tagCount, count(*) as "Artistcount" from ( select count(*) as "tagCount" from artist_tag group by artist_id) a group by a.tagcount order by 2 desc '
pd = (sql.read_sql_query(q1, cnx))
pd.head(25)


tagCount  Artistcount
0          1          414
1          2          183
2          3          110
3          4           96
4          5           75
5          7           48
6          6           40
7          8           26
8          9           24
9         10           23
10        11           14
11        12            8
12        14            8
13        13            6
14        16            6
15        15            4
16        20            4
17        17            2

now, select all tags for one artist, also join on artist_id_name to get artist name (use artist_id from above, one with 20 tags)

In [12]:
q1 = 'select a.artist_id, b.artist_name ,a.mbtag from artist_tag a inner join artist_id_name b on a.artist_id = b.artist_id where a.artist_id = "AR4L4WQ1187FB51996" '
pd = (sql.read_sql_query(q1, cnx))
pd.head(20)

artist_id  artist_name                  mbtag
0   AR4L4WQ1187FB51996  The Prodigy             electronic
1   AR4L4WQ1187FB51996  The Prodigy                british
2   AR4L4WQ1187FB51996  The Prodigy               big beat
3   AR4L4WQ1187FB51996  The Prodigy                 techno
4   AR4L4WQ1187FB51996  The Prodigy               hardcore
5   AR4L4WQ1187FB51996  The Prodigy                     uk
6   AR4L4WQ1187FB51996  The Prodigy            electropunk
7   AR4L4WQ1187FB51996  The Prodigy              braintree
8   AR4L4WQ1187FB51996  The Prodigy  dance and electronica
9   AR4L4WQ1187FB51996  The Prodigy     breakbeat hardcore
10  AR4L4WQ1187FB51996  The Prodigy             dance-punk
11  AR4L4WQ1187FB51996  The Prodigy                electro
12  AR4L4WQ1187FB51996  The Prodigy                  dance
13  AR4L4WQ1187FB51996  The Prodigy                   2005
14  AR4L4WQ1187FB51996  The Prodigy            britannique
15  AR4L4WQ1187FB51996  The Prodigy            electronica
16  AR4L4WQ1187FB51996  The Prodigy                english
17  AR4L4WQ1187FB51996  The Prodigy       alternative rock
18  AR4L4WQ1187FB51996  The Prodigy                   rave
19  AR4L4WQ1187FB51996  The Prodigy             industrial

Interesting, the artists 'The Prodigy' have an extensive list of tags. 

Now, check how many occurances there are for each of the prodigy tags ( check how connected they could be)

In [13]:
q1 = 'select mbtag, count(*) as "count" from artist_tag  where mbtag in ( select mbtag from artist_tag where artist_id = "AR4L4WQ1187FB51996" ) group by mbtag order by 2 desc '
pd = (sql.read_sql_query(q1, cnx))
pd.head(20)

mbtag  count
0                 british    122
1                      uk    119
2                 english     87
3        alternative rock     52
4              electronic     50
5   dance and electronica     24
6             britannique     21
7             electronica     14
8                   dance      9
9                  techno      9
10                electro      6
11               hardcore      5
12             industrial      4
13             dance-punk      3
14               big beat      2
15            electropunk      2
16                   rave      2
17                   2005      1
18              braintree      1
19     breakbeat hardcore      1

This looks good, there are a lot of connections 

Now, get the avg number of tags for an artist

In [14]:
q1 = 'select avg(a.count) from (select count(*) as "count" from artist_id_name group by artist_id) a '
pd = (sql.read_sql_query(q1, cnx))
pd.head()

avg(a.count)
0      1.141204

So, the avg tags for most artists will be about 1, thats ok 

####### Now, check if any artists with Tags have no songs OR artists with songs have no tags 

####### Summary of below

All artists on the artist_tag table have a song on the songs table, which is grand.
There are only about 1090 unique artists on artist_tag table so about 3060 artists on the songs table are without any tags 


In [36]:
# select a count of artists on the artist_tags table that do not have a song (check against artist_id_name table)
q1='select "No of Artists on Tags w/o songs", count(*) as "count" from artist_tag where artist_id not in (select artist_id from artist_id_name)'
pd = (sql.read_sql_query(q1, cnx))
pd.head(10) 

"No of Artists on Tags w/o songs"  count
0   No of Artists on Tags w/o songs      0

In [40]:
# select a count of artists with a song but no tag on the artist_tag (check using artist_id_name table)
q1='select "No of Artists on Songs w/o tags", count(*) as "count" from artist_id_name where artist_id not in (select artist_id from artist_tag)'
pd = (sql.read_sql_query(q1, cnx))
pd.head(10) 

"No of Artists on Songs w/o tags"  count
0   No of Artists on Songs w/o tags   3106

In [41]:
# im not sure about this must do a count of distinct artist_ids on on the artist_tag table, if there are
# only around 1000 or so, then its ok
q1='select count(distinct artist_id ) from artist_tag'
pd = (sql.read_sql_query(q1, cnx))
pd.head(10) 

count(distinct artist_id )
0                        1091

Now, go to the artist_artist_similarty table and get a count of of artists that are similar to an artist (first select from table)

In [27]:
q1 = 'select * from artist_artist_similarty limit 10'
pd = (sql.read_sql_query(q1, cnx))
pd.head(10)

index              target             similar
0      0  AR009211187B989185  ARHINI31187B995C1D
1      1  AR009211187B989185  ARI0PUX1187FB3F215
2      2  AR009211187B989185  AR9RTS51187B996CC8
3      3  AR009211187B989185  ARS1DCR1187B9A4A56
4      4  AR009211187B989185  ARTPGR61187B98B0F6
5      5  AR009211187B989185  ARIYH1U1187B9AF8A3
6      6  AR009211187B989185  ARHB1961187B9B1732
7      7  AR009211187B989185  ARI4SKQ1187B9B43B0
8      8  AR009211187B989185  AR3OSJM1187B98B95B
9      9  AR009211187B989185  ARIRD6J1187FB5A98C

the above shows the target (artist) and similar artists
Now, get a count of similar artists for an artists 

In [16]:
q1 = 'select a.target, count(*) as "count" from artist_artist_similarty a group by a.target order by 2 desc '
pd = (sql.read_sql_query(q1, cnx))
pd.head(10)

target  count
0  ARB1LM61187FB4DDF4     47
1  ARF3XJD1187FB38DE0     47
2  ARRZFUG1187B98AD42     46
3  ARGUACZ1187FB3F35C     45
4  AR95F611187B99CE1C     44
5  AREB9421187FB458EF     43
6  ARF9BNW1187FB3F54B     43
7  ARGHPIE1187FB3CCDF     40
8  AR6ZYK41187B9996BB     39
9  ARC28SN1187B989049     39

There are many connections, select all similar artists for one artist, also join on artist_id_name to get artist name (use artist_id 'target' from above, one with 45 similar artists)

In [17]:
q1 = 'select b.artist_name, c.artist_name from artist_artist_similarty a inner join artist_id_name b on a.target = b.artist_id  inner join artist_id_name c on a.similar = c.artist_id  where a.target = "ARGUACZ1187FB3F35C" '
pd = (sql.read_sql_query(q1, cnx))
pd.head(20)

artist_name                  artist_name
0   Connie Smith                Tammy Wynette
1   Connie Smith                 Hank Locklin
2   Connie Smith                    Ray Price
3   Connie Smith               Porter Wagoner
4   Connie Smith                  Sonny James
5   Connie Smith                  Dottie West
6   Connie Smith                  Kitty Wells
7   Connie Smith                   Don Gibson
8   Connie Smith                  Donna Fargo
9   Connie Smith                 Loretta Lynn
10  Connie Smith                  Faron Young
11  Connie Smith                Charley Pride
12  Connie Smith                Skeeter Davis
13  Connie Smith  PATSY CLINE & WILLIE NELSON
14  Connie Smith                  Patsy Cline
15  Connie Smith                   Roy Drusky
16  Connie Smith                 Dolly Parton
17  Connie Smith                 Charlie Rich
18  Connie Smith            Barbara Fairchild
19  Connie Smith                Jeanne Pruett

The artist - artist connections are looking good
Now, get the avg number of similar artists for an artist

In [43]:
q1 = 'select avg(a.count) from (select count(*) as "count" from artist_artist_similarty group by target) a '
pd = (sql.read_sql_query(q1, cnx))
pd.head()

avg(a.count)
0     11.409719

An avg of 11 similar artists for a given artist is plenty for connections  

####### Now, check if any artists with similar artists have no songs OR artists with songs have enteries on the similar table
(note : As both columns on the artist_artist_similar table as artist_ID values (target and similar), i check both on the artist
table. I did this in one sql statement using a join, just easier to see results together.

####### Summary of below

All artists on the artist_artist_similar table have a song on the songs table, there are about 3700 of these, good.
There are about about 530 artists on the songs table are without a target reference on the artist_artist_similarty table

In [47]:
# select a count of artists on the artist_artist_similar table that do not have a song, use union (check against artist_id_name table)
q1='select "Num of Artists in similar w/o songs", count(*) as "count" from artist_artist_similarty where target not in (select artist_id from artist_id_name) union select "Num of similar Artists in similar w/o songs", count(*) as "count" from artist_artist_similarty where similar not in (select artist_id from artist_id_name) '
pd = (sql.read_sql_query(q1, cnx))
pd.head(10) 

"Num of Artists in similar w/o songs"  count
0          Num of Artists in similar w/o songs      0
1  Num of matched Artists in similar w/o songs      0

In [51]:
# select a count of artists on the artist_artist_similar table that do not have a song, use union (check against artist_id_name table)
q1='select "No of Artists on Songs w/o similar", count(*) as "count" from artist_id_name where artist_id not in (select target from artist_artist_similarty) union select "No of Artists on Songs w/o matched Artists", count(*) as "count" from artist_id_name where artist_id not in (select similar from artist_artist_similarty) '
pd = (sql.read_sql_query(q1, cnx))
pd.head(10) 

"No of Artists on Songs w/o similar"  count
0  No of Artists on Songs w/o matched Artists    532
1          No of Artists on Songs w/o similar    127

In [53]:
# do a count of distinct targets (artist_ids) on the artist_artist_similar table 
q1='select count(distinct target ) from artist_artist_similarty'
pd = (sql.read_sql_query(q1, cnx))
pd.head(10) 

count(distinct target )
0                     3766

This is sampe update code, it can be used for any data cleaning

1. Select Data

    query1 = " select * from artist_tag where artist_id = 'ARZYP6O1187B9892E7'  " 
    
    pd = (sql.read_sql_query(query1, cnx))
    
    print(pd.head(10))

2. Update ( cursor, update, commit) 


    c = cnx.cursor()
    
    c.execute("update artist_tag set mbtag = 'test new'  where artist_id = 'ARZYP6O1187B9892E7' " ) 
    
    cnx.commit()
    
        
3. Select Data, confirm change

    query1 = " select * from artist_tag where artist_id = 'ARZYP6O1187B9892E7'  " 
    
    pd = (sql.read_sql_query(query1, cnx))
    
    print(pd.head(10))